TIM/TIM Core Code/man/man3/buffer.3
.TH "buffer" 3 "Tue Jan 26 2016" "Version 0.1" "P21451_TIM" \" -*- nroff -*-
.ad l
.nh
.SH NAME
buffer \- FIFO Byte Buffer
.SS "Modules"

.in +1c
.ti -1c
.RI "\fBprintf Functionality for Buffer Module\fP"
.br
.in -1c
.SS "Data Structures"

.in +1c
.ti -1c
.RI "struct \fBbuffer_t\fP"
.br
.in -1c
.SS "Macros"

.in +1c
.ti -1c
.RI "#define \fBBUFFER_PUSH_FAILED\fP   1"
.br
.RI "\fIpush failed (return value of \fBPushData()\fP ) \fP"
.ti -1c
.RI "#define \fBBUFFER_PUSH_SUCCEEDED\fP   0"
.br
.RI "\fIpush succeeded (return value of \fBPushData()\fP ) \fP"
.in -1c
.SS "Functions"

.in +1c
.ti -1c
.RI "void \fBPush\fP (\fBbuffer_t\fP *buffer, char data)"
.br
.ti -1c
.RI "char \fBPop\fP (\fBbuffer_t\fP *buffer)"
.br
.ti -1c
.RI "uint16_t \fBGetSize\fP (\fBbuffer_t\fP *buffer)"
.br
.ti -1c
.RI "void \fBBufferInit\fP (\fBbuffer_t\fP *buffer, char *data_array, uint16_t max_size)"
.br
.ti -1c
.RI "void \fBBufferSetCallback\fP (\fBbuffer_t\fP *buffer, void(*Callback)(\fBbuffer_t\fP *buffer))"
.br
.ti -1c
.RI "void \fBBufferClearCallback\fP (\fBbuffer_t\fP *buffer)"
.br
.ti -1c
.RI "char \fBPushData\fP (\fBbuffer_t\fP *buffer, char *data, uint16_t length)"
.br
.in -1c
.SH "Detailed Description"
.PP 
This module implements a software FIFO buffer of bytes\&. It provides methods \fBPush()\fP and \fBPop()\fP to add and remove bytes from the buffer\&.
.PP
The user is responsible for allocating the structure used to manage the buffer, \fBbuffer_t\fP, as well as the actual byte array which will be used to implement the buffer\&. These are then passed to \fBBufferInit()\fP which must be called prior to any attempts to Push or Pop\&.
.PP
There is also a \fBPushData()\fP method which allows multiple bytes to be added to the buffer at once\&.
.PP
An alternative to this FIFO byte buffer is the Item Buffer module which works the same except with items of any size instead of being limited to bytes only\&.
.PP
\fBAuthor:\fP
.RS 4
Michael Muhlbaier 
.PP
Anthony Merlino
.RE
.PP
\fBVersion:\fP
.RS 4
0\&.1 Initial implementation 
.PP
1\&.0 Made Push and Pop interrupt safe 
.PP
1\&.1 Added PushData 
.PP
1\&.2 Added and corrected documentation, removed include 'system\&.h' 
.RE
.PP

.SH "Macro Definition Documentation"
.PP 
.SS "#define BUFFER_PUSH_FAILED   1"

.PP
push failed (return value of \fBPushData()\fP ) 
.PP
Definition at line 189 of file buffer\&.h\&.
.SS "#define BUFFER_PUSH_SUCCEEDED   0"

.PP
push succeeded (return value of \fBPushData()\fP ) 
.PP
Definition at line 190 of file buffer\&.h\&.
.SH "Function Documentation"
.PP 
.SS "void BufferClearCallback (\fBbuffer_t\fP * buffer)"
Clear/remove the callback function for 'buffer'
.PP
\fBParameters:\fP
.RS 4
\fIbuffer\fP Pointer to the \fBbuffer_t\fP data structure whose callback function is to be cleared 
.RE
.PP

.PP
Definition at line 83 of file buffer\&.c\&.
.SS "void BufferInit (\fBbuffer_t\fP * buffer, char * data_array, uint16_t max_size)"
Initialize a FIFO buffer
.PP
Example code: 
.PP
.nf
1 #define TX_BUFFER_LENGTH 512
2 buffer_t tx; // transmit buffer
3 char tx_buffer_array[TX_BUFFER_LENGTH]
4 \&.\&.\&.
5 BufferInit(&tx, &tx_buffer_array[0], TX_BUFFER_LENGTH);

.fi
.PP
.PP
\fBParameters:\fP
.RS 4
\fIbuffer\fP Pointer to the \fBbuffer_t\fP data structure to be initialized 
.br
\fIdata_array\fP Array of char data to implement the actual buffer 
.br
\fImax_size\fP Maximum size of the buffer (should be the same length as the array) 
.RE
.PP

.PP
Definition at line 67 of file buffer\&.c\&.
.SS "void BufferSetCallback (\fBbuffer_t\fP * buffer, void(*)(\fBbuffer_t\fP *buffer) Callback)"
Set Callback function for buffer to be called after items are Push'd to the buffer
.PP
The callback function will be called after anything is Push'd to the buffer\&. The function will be called with a pointer to the buffer which had an item pushed onto it\&.
.PP
Example: 
.PP
.nf
1 void TxCallback(buffer_t * buf);
2 #define TX_BUFFER_LENGTH 512
3 buffer_t tx; // transmit buffer
4 char tx_buffer_array[TX_BUFFER_LENGTH]
5 \&.\&.\&.
6 BufferInit(&tx, &tx_buffer_array[0], TX_BUFFER_LENGTH);
7 BufferSetCallback(&tx, TxCallback);
8 \&.\&.\&.
9 void TxCallback(buffer_t * buf) {
10         SET_UART_TX_IE();
11 }

.fi
.PP
 This example is useful for a uC which has a hardware Tx interrupt flag which is set whenever there is room in the hardware Tx FIFO buffer\&. When done transmitting the interrupt must be disabled to avoid getting stuck in the ISR\&. When data needs to be sent the interrupt must be enabled again, thus the need for the callback\&.
.PP
Another usage could be to handle received data on a receive buffer\&.
.PP
\fBWarning:\fP
.RS 4
many applications may use the Push method in a ISR which means the Callback would be called from a ISR\&. Thus care should be taken to ensure callbacks are both fast and interrupt safe
.RE
.PP
\fBParameters:\fP
.RS 4
\fIbuffer\fP Pointer to the \fBbuffer_t\fP data structure whose callback function is to be set 
.br
\fICallback\fP Function pointer to a callback function with no return value and a \fBbuffer_t\fP pointer input\&. 
.RE
.PP

.PP
Definition at line 76 of file buffer\&.c\&.
.SS "uint16_t GetSize (\fBbuffer_t\fP * buffer)"
GetSize returns the number of items in the FIFO buffer
.PP
\fBBufferInit()\fP should be used to initialize the buffer otherwise the return value will be meaningless
.PP
\fBParameters:\fP
.RS 4
\fIbuffer\fP Pointer to the \fBbuffer_t\fP data structure holding the buffer info 
.RE
.PP
\fBReturns:\fP
.RS 4
Number of items in the buffer 
.RE
.PP

.PP
Definition at line 63 of file buffer\&.c\&.
.SS "char Pop (\fBbuffer_t\fP * buffer)"
Pop will return one item from the front of the FIFO buffer
.PP
Pop will return the item at the front of the FIFO buffer then increment (and wrap as needed) the front\&. If the buffer is empty it will return 0\&.
.PP
\fBBufferInit()\fP must be used to initialize the buffer prior to calling Pop and passing it a pointer to the buffer\&.
.PP
\fBParameters:\fP
.RS 4
\fIbuffer\fP Pointer to the \fBbuffer_t\fP data structure holding the buffer info 
.RE
.PP
\fBReturns:\fP
.RS 4
Data of type char from the front of the buffer
.RE
.PP
\fBWarning:\fP
.RS 4
is only interrupt safe if \fBEnableInterrupts()\fP and \fBDisableInterrupts()\fP are defined by \fBhal_general\&.h\fP 
.RE
.PP

.PP
Definition at line 47 of file buffer\&.c\&.
.SS "void Push (\fBbuffer_t\fP * buffer, char data)"
Push will add one item, data, to the FIFO buffer
.PP
Push will add one item to the rear of the data buffer then increment (and wrap is needed) the rear\&. If the buffer is full it will overwrite the data at the front of the buffer and increment the front\&.
.PP
\fBBufferInit()\fP must be used to initialize the buffer prior to calling Push and passing it a pointer to the buffer\&.
.PP
\fBParameters:\fP
.RS 4
\fIbuffer\fP Pointer to the \fBbuffer_t\fP data structure holding the buffer info 
.br
\fIdata\fP char data to be added to the rear of the FIFO buffer
.RE
.PP
\fBWarning:\fP
.RS 4
Push is only interrupt safe if \fBEnableInterrupts()\fP and \fBDisableInterrupts()\fP are defined by \fBhal_general\&.h\fP 
.RE
.PP

.PP
Definition at line 19 of file buffer\&.c\&.
.SS "char PushData (\fBbuffer_t\fP * buffer, char * data, uint16_t length)"
Push a array of data to the buffer
.PP
\fBWarning:\fP
.RS 4
PushData will disable interrupts while writing to the buffer\&. If length is long this could interfere with time sensitive ISRs\&. Consider using \fBPush()\fP if this is an issue\&.
.RE
.PP
PushData will add an array of items to the rear of the data buffer and increment (and wrap if needed) the rear\&. If the buffer does not have room none of the data will be pushed to the buffer\&.
.PP
\fBBufferInit()\fP must be used to initialize the buffer prior to calling PushData and passing it a pointer to the buffer\&.
.PP
\fBParameters:\fP
.RS 4
\fIbuffer\fP Pointer to the \fBbuffer_t\fP data structure holding the buffer info 
.br
\fIdata\fP char pointer to data array to be added to the rear of the FIFO buffer 
.br
\fIlength\fP the number of items in the data array 
.RE
.PP
\fBReturns:\fP
.RS 4
0 if succeeded, 1 if no room in buffer 
.RE
.PP

.PP
Definition at line 85 of file buffer\&.c\&.
.SH "Author"
.PP 
Generated automatically by Doxygen for P21451_TIM from the source code\&.


TIM/TIM Core Code/buffer.c
//x comments which begin with //x are for beginner student information only
//x these comments would normally not be required
#include "system.h"
//x hal_general includes general Hardware Abstraction Layer definitions
//x in this module those are BlockInterrupts() and RestoreInterrupts()
//x these two functions are abstracted from here since they are implemented
//x differently on different hardware
#include "hal_general.h"

#ifndef _BUFFER_H_
#error "buffer.h must be included in system.h"
#endif
//x comments in your C file are only nedded to explain more complicated
//x implementation techniques. You don't need to describe things that are
//x plainly apparent. My general rule is "would future you benefit from a
//x comment explaining what you are implementing"
//x the documentation for a function which is "public" (in the .h file) is
//x located in the .h file
void Push(buffer_t *buffer, char data) {
	// disable interrupts to make the function interrupt safe
	BlockInterrupts();
	
	/*	If front = rear and the buffer is not empty then overwrite what is on
		the front, this allows use as a revolving circular buffer for receivers */
	if(buffer->front == buffer->rear && buffer->size) {
		buffer->front++;
		// check to make sure front isn't past the end of the buffer
		if(buffer->front > buffer->buffer_end)
			buffer->front = buffer->buffer_start;
		buffer->size--;
	}
	
	// push data onto rear location and increment rear pointer
	*buffer->rear++ = data;
	// check to make sure rear isn't past the end of the buffer
	if(buffer->rear > buffer->buffer_end) buffer->rear = buffer->buffer_start;
	// Increment the size of the buffer
	buffer->size++;

	// If there is a callback, call it
	if(buffer->Callback) buffer->Callback(buffer);

	//ReEnable Interrupts
	RestoreInterrupts();
}

char Pop(buffer_t *buffer) {
	char data;
	//x in light of what was documented in the Push function in this file
	//x there is nothing in the Pop function which is not plain given the
	//x good use of variable names in this module
	BlockInterrupts();
	data = *buffer->front;
	if(buffer->size != 0) {
		buffer->front++;
		if(buffer->front > buffer->buffer_end) buffer->front = buffer->buffer_start;
		buffer->size--;
	}
	RestoreInterrupts();
	return data;
}

uint16_t GetSize(buffer_t *buffer) {
	return buffer->size;
}

void BufferInit(buffer_t *buffer, char *data_array, uint16_t max_size) {
	buffer->buffer_start = data_array;
	buffer->front = data_array;
	buffer->rear = data_array;
	buffer->buffer_end = data_array + max_size - 1;
	buffer->size = 0;
	buffer->Callback = 0;
}

void BufferSetCallback(buffer_t * buffer, void (*Callback)(buffer_t * buffer)) {
	buffer->Callback = (void (*)(void *))Callback; // cast callback to void pointer
	//x a void pointer was used in the structure for buffer_t becuase a buffer_t pointer
	//x couldn't be used since buffer_t hadn't been defined inside of the definition of
	//x buffer_t. See the .h file for more info.
}

void BufferClearCallback(buffer_t * buffer) { buffer->Callback = 0; }

char PushData(buffer_t * buffer, char * data, uint16_t length) {
	BlockInterrupts();
	// check if there is enough space in the buffer to load the data
	if(buffer->buffer_end - buffer->buffer_start + 1 - buffer->size >= length) {
		buffer->size += length;
		while(length) {
			// push data onto rear location and increment rear
			*buffer->rear++ = *data++;
			// check to make sure rear isn't past the end of the buffer
			if(buffer->rear > buffer->buffer_end) buffer->rear = buffer->buffer_start;
			length--;
		}
		if(buffer->Callback) buffer->Callback(buffer);

		RestoreInterrupts();

		return BUFFER_PUSH_SUCCEEDED;
	}

	RestoreInterrupts();

	return BUFFER_PUSH_FAILED;
}

